src/pipeline/mod.rs

```diff
@@ -0,0 +1,39 @@
+use crate::error::Res;
+
+pub trait PStep {
+    type Input;
+    type Output;
+    /// takes inputs from previous step, transforms to produce outputs for next step.
+    /// # Errors
+    /// If there was an unrecoverable error when processing inputs, return error and abort
+    /// computation.
+    fn compute(&self, inp: Self::Input) -> Res<Self::Output>;
+}
+
+/// shortcut to composing all of the steps of the pipeline.
+/// takes a list of `PStep`s and feeds the outputs of one into the inputs of the next
+/// final result is a function with arguments matching `Input` of the first `PStep`, and return
+/// value matching `Res<Output>` of the last `PStep`.
+#[macro_export]
+macro_rules! build_pipeline {
+    ($last:expr) => {{
+        let last = ($last);
+        move |x| last.compute(x)
+    }};
+    ($head:expr, $($tail:expr),+) => {{
+        let head = ($head);
+        move |x| {
+            let outer_res = head.compute(x)?;
+            let inner = build_pipeline!($($tail),+);
+            let inner_res = inner(outer_res);
+            inner_res
+        }
+    }};
+}
+
+pub trait Pipeline<Input, Output> {
```

Comment:
I was thinking this trait would be good to be able to define different pipelines that are composable at runtime.

e.g. you'd have a "secure mul" pipeline, a "shuffle" pipeline, which you can string together to build a protocol. and in that sense, I thought it'd be good to give these pieces names and structures.

One other thought I had was maybe the macro would literally produce a `Pipeline`, or maybe a `PStep`, instead of a function. that keeps it composable